File: ABTO_iPhone_newVoipSDK_trial_x64(20160923)/Sample_1/Sample_1/UI/MainPageViewController.m
//
//  MainPageViewController.m
//
//  Created on 12/8/11.
//  Copyright (c) 2011 ABTO Software. All rights reserved.
//

#import "MainPageViewController.h"
#import "SettingsViewController.h"
#import "DialerViewController.h"
#import "MainTabBarViewController.h"
#import "VideoCallViewController.h"

@implementation MainPageViewController

- (void)saveChanges {
    if (name.text == nil) {
        name.text = @"";
    }
    if (password.text == nil) {
        password.text = @"";
    }
    if (domain.text == nil) {
        domain.text = @"";
    }
    AbtoPhoneConfig* config = abtoAppDelegate.sharedInstance.phone.config;
    config.regUser = name.text;
    config.regPassword = password.text;
    config.regDomain = domain.text;
    // for bundled resources use relative path, for global resources URL
    //config.ringToneUrl = @"abto_ringtone.mp3";
    //config.ringBackToneUrl = @"abto_ringbacktone.wav";
    [config saveToUserDefaults:SETTINGS_KEY];
}

- (IBAction)enterClick:(id)sender {
    if ((domain.text && ![domain.text isEqualToString:@""]) &&
        (name.text && ![name.text isEqualToString:@""]) &&
        (password.text && ![password.text isEqualToString:@""])){

        [self saveChanges];

        if ([[abtoAppDelegate sharedInstance].phone finalizeConfiguration]) {
            [activityIndicator startAnimating];
            buttonEnter.enabled = NO;
        }
    } else {
        UIAlertView *requiredFieldsAreEmpty = [[UIAlertView alloc] initWithTitle:@"Required fields are empty" message:@"Please fill in all the fields" delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles: nil];
        [requiredFieldsAreEmpty show];
    }
}

- (IBAction)settingsClick:(id)sender {
    [self saveChanges];
    SettingsViewController *nextController = [SettingsViewController new];
    [nextController.navigationController setNavigationBarHidden:NO animated:NO];
    UINavigationController *navigationController = [[UINavigationController alloc] initWithRootViewController:nextController];
    
    [ApplicationDelegate.navController presentViewController:navigationController animated:YES completion:nil];
}

- (IBAction)textFieldDoneEditing:(id)sender {
    if ((sender == name) || (sender == password) || (sender == domain)) {
        [sender becomeFirstResponder];
    }
}

- (IBAction)backgroundClick:(id)sender {
    [name resignFirstResponder];
    [password resignFirstResponder];
    [domain resignFirstResponder];
}

-(void)refreshFields {
    AbtoPhoneConfig* config = [abtoAppDelegate sharedInstance].phone.config;
    name.text = config.regUser;
    password.text = config.regPassword;
    domain.text = config.regDomain;
}

#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    CGRect screen = [[UIScreen mainScreen] bounds];
    CGRect rect = [buttonEnter frame];

    rect.origin.y = screen.size.height * 0.8f;
    [buttonEnter setFrame:rect];

    rect = [name frame];
    rect.origin.y = screen.size.height * 0.3f;
    [name setFrame:rect];
    
    rect = [password frame];
    rect.origin.y = [name frame].origin.y + [name frame].size.height + screen.size.height * 0.05f;
    [password setFrame:rect];
    
    rect = [domain frame];
    rect.origin.y = [password frame].origin.y + [password frame].size.height + screen.size.height * 0.05f;
    [domain setFrame:rect];
    
    [self refreshFields];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onPhoneNotification:) name:NOTIFICATION_PHONE_EVENT object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onCallNotification:) name:NOTIFICATION_CALL_EVENT object:nil];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    buttonEnter.enabled = YES;
    [activityIndicator stopAnimating];
}

-(void)viewDidAppear:(BOOL)animated {
    [self refreshFields];
	[super viewDidAppear:animated];    
}

-(void)viewWillDisappear:(BOOL)animated {
    [self saveChanges];
    [super viewWillDisappear:animated];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)onPhoneNotification:(NSNotification *)notification {
    PhoneEvents status = [[notification object] integerValue];
    BOOL presentInStack = NO;
    switch (status) {
        case PhoneEventsRemoteAlerting:
            [activityIndicator startAnimating];
            DLog(@"onPhoneNotification Remote Alerting");
            break;
        case PhoneEventsRegSuccess:
            [activityIndicator stopAnimating];
            ApplicationDelegate.connected = YES;
            DLog(@"[ApplicationDelegate.navController.topViewController class] = %@", [ApplicationDelegate.navController.topViewController class]);
            for (UIViewController* controller in ApplicationDelegate.navController.viewControllers) {
                if ([controller isKindOfClass:[MainTabBarViewController class]]) {
                    presentInStack = YES;
                    break;
                }
            }
            
            if (!presentInStack){
                [ApplicationDelegate.navController pushViewController:[MainTabBarViewController new] animated:YES];
                DLog(@"View was pushed again");
            }
            DLog(@"onPhoneNotification Registered");
            break;
        case PhoneEventsUnregSuccess:
            [activityIndicator stopAnimating];
            buttonEnter.enabled = YES;
            DLog(@"onPhoneNotification Unregistered");
            break;
        case PhoneEventsUnregFailed:
            // FIX:just to make sure to logout in case we failed prolongating expiration time on SIP server
            //[abtoAppDelegate logout];
            DLog(@"onPhoneNotification Register Failed");
        default:
            [activityIndicator stopAnimating];
            buttonEnter.enabled = YES;
            break;
    }
}

- (void)onCallNotification:(NSNotification *)notification {
    NSInteger status = [[notification object] integerValue];
    
    if (status == CallEventsIncoming) {
        NSInteger callId = [[notification.userInfo valueForKey:CALL_ID_ARGUMENT] integerValue];
        NSString *contact = [notification.userInfo valueForKey:CONTACT_ARGUMENT];
        
        if (callId != kInvalidCallId) {
            UIApplication *app = [UIApplication sharedApplication];
            abtoAppDelegate *appDelegate = [abtoAppDelegate sharedInstance];
            NSString *remotePartyNumber = [AbtoPhoneInterface sipUriUsername:contact];
            UIApplicationState state = app.applicationState;
            
            [appDelegate.phone setSpeakerphoneOn:YES];
            if ((state == UIApplicationStateBackground) || (state == UIApplicationStateInactive)) {
                UILocalNotification *localNotification = [UILocalNotification new];
                if (localNotification) {
                    localNotification.alertBody = [NSString stringWithFormat:@"%@ %@", remotePartyNumber, @"calling"];
                    //            localNotif.soundName = UILocalNotificationDefaultSoundName;
                    localNotification.applicationIconBadgeNumber = ++app.applicationIconBadgeNumber;
                    localNotification.repeatInterval = 0;
                    NSDictionary *userInfo = @{ kNotifKey: kNotifKey_IncomingCall,
                                                kNotifCall_SessionId : @(callId),
                                                kNotifCall_RemoteInfo: contact };

                    localNotification.userInfo = userInfo;
                    [[UIApplication sharedApplication]  presentLocalNotificationNow:localNotification];
                    ApplicationDelegate.lastCallNotification = localNotification;
                }
                
            } else {
                CallViewController *nextController = [appDelegate.phone isVideoCall:callId] ? [VideoCallViewController new] : [AudioCallViewController new];
                nextController.number = remotePartyNumber;
                nextController.callId = callId;
                nextController.incoming = YES;
                [appDelegate.navController pushViewController:nextController animated:YES];
            }
        }
    } else if (status == CallEventsDisconnected) {
        ApplicationDelegate.lastCallNotification = nil;
    }
}

@end

File: ABTO_iPhone_newVoipSDK_trial_x64(20160923)/SampleSwift/SampleSwift/MainPageViewController.swift
//
//  MainPageViewController.swift
//  SampleSwift
//
//  Created by Nazariy Vlizlo on 10/19/15.
//  Copyright © 2015 ABTO Software. All rights reserved.
//

import UIKit

class MainPageViewController: UIViewController, UITextFieldDelegate {
    
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    @IBOutlet weak var name: UITextField!
    @IBOutlet weak var password: UITextField!
    @IBOutlet weak var domain: UITextField!
    @IBOutlet weak var buttonEnter: UIButton!
    
    //MARK:ViewController Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        refreshFields()

        NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(MainPageViewController.onCallNotification(_:)), name: NOTIFICATION_CALL_EVENT, object: nil)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.setNavigationBarHidden(true, animated: false)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        refreshFields()
        activityIndicator.stopAnimating()
        NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(MainPageViewController.onConnectionStatusNotification(_:)), name: NOTIFICATION_PHONE_EVENT, object: nil)
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        navigationController?.setNavigationBarHidden(false, animated: false)
    }
    
    override func shouldAutorotate() -> Bool {
        let orientation = UIApplication.sharedApplication().statusBarOrientation
        return orientation == .Portrait
    }
    
    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        name.resignFirstResponder()
        password.resignFirstResponder()
        domain.resignFirstResponder()
    }
    
    //MARK:Custom methods
    
    func saveChanges() {
        let config = AppDelegate.sharedInstance().phone!.config()
        if let abtoPhoneConfig = config {
            abtoPhoneConfig.regUser = name.text
            abtoPhoneConfig.regPassword = password.text
            abtoPhoneConfig.regDomain = domain.text
            abtoPhoneConfig.enableRingBackTone = true
            abtoPhoneConfig.enableRingTone = true
            abtoPhoneConfig.regExpirationTime = 10
            //abtoPhoneConfig.ringBackToneUrl = "abto_ringbacktone.wav"
            //abtoPhoneConfig.ringToneUrl = "abto_ringtone.mp3"
            abtoPhoneConfig.saveToUserDefaults(SETTINGS_KEY)
        }
    }
    
    func refreshFields() {
        let config = AppDelegate.sharedInstance().phone?.config()
        name.text = config?.regUser
        password.text = config?.regPassword
        domain.text = config?.regDomain
    }
    
    func onCallNotification(notification: NSNotification) {
        let status = CallEvents(rawValue: notification.object as! Int)
        if status == .Incoming {
            let callId = notification.userInfo![CALL_ID_ARGUMENT] as? Int
            let contact = notification.userInfo![CONTACT_ARGUMENT] as? String
            
            if callId != kInvalidCallId {
                let app = UIApplication.sharedApplication()
                let appDelegate = AppDelegate.sharedInstance()
                let remotePartyNumber = AbtoPhoneInterface.sipUriUsername(contact)
                let applicationState = app.applicationState
                appDelegate.phone?.setSpeakerphoneOn(true)
                if applicationState == .Background || applicationState == .Inactive {
                    let localNotification = UILocalNotification()
                    localNotification.alertBody = "\(remotePartyNumber) calling"
                    app.applicationIconBadgeNumber += 1
                    localNotification.applicationIconBadgeNumber = app.applicationIconBadgeNumber
                    localNotification.repeatInterval = NSCalendarUnit(rawValue: 0)
                    let userInfo: [String: AnyObject] = [
                        kNotifKey: kNotifKey_IncomingCall,
                        kNotifCall_SessionId: callId!,
                        kNotifCall_RemoteInfo: contact!
                    ]
                    localNotification.userInfo = userInfo
                    UIApplication.sharedApplication().presentLocalNotificationNow(localNotification)
                    applicationDelegate().lastCallNotification = localNotification
                } else {
                    let nextController = storyboard?.instantiateViewControllerWithIdentifier(appDelegate.phone!.isVideoCall(callId!) ? "VideoCallVC" : "AudioCallVC") as? CallViewController
                    nextController?.number = remotePartyNumber
                    nextController?.callId = callId
                    nextController?.incoming = true
                    navigationController?.pushViewController(nextController!, animated: true)
                }
            }
        } else if status == .Disconnected {
            applicationDelegate().lastCallNotification = nil
        }
    }
    
    func onConnectionStatusNotification(notification: NSNotification) {
        let status = PhoneEvents(rawValue: notification.object!.integerValue)!
        switch status {
        case .RemoteAlerting:
            activityIndicator.startAnimating()
            print("onConnectionStatusNotification ClientConnectionStatusRegistering")
        case .RegSuccess:
            NSNotificationCenter.defaultCenter().removeObserver(self, name: NOTIFICATION_PHONE_EVENT, object: nil)
            activityIndicator.stopAnimating()
            let tabbarVC = storyboard?.instantiateViewControllerWithIdentifier("MainTabBar") as! MainTabBarController
            
            navigationController?.pushViewController(tabbarVC, animated: true)
        case .UnregSuccess:
            activityIndicator.stopAnimating()
            print("onConnectionStatusNotification ClientConnectionStatusUnregistered")
        case .UnregFailed:
            print("onConnectionStatusNotification ClientConnectionStatusDisconnected")
        default:
            activityIndicator.stopAnimating()
        }
    }
    
    //MARK:Actions
    
    @IBAction func enterClick(sender: UIButton) {
        if (domain.text?.characters.count > 0) &&
            (name.text?.characters.count > 0) &&
            (password.text?.characters.count > 0) {
                saveChanges()
                
                if AppDelegate.sharedInstance().phone?.finalizeConfiguration() == true {
                    activityIndicator.startAnimating()
                }
        } else {
            let requiredFieldsAreEmpty = UIAlertView(title: "Required fields are empty", message: "Please fill in all the fields", delegate: nil, cancelButtonTitle: "OK")
            requiredFieldsAreEmpty.show()
        }
    }
    
    @IBAction func settingsClick(sender: UIButton) {
        saveChanges()
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        return textField.resignFirstResponder()
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
}
